Improve performance and add support for versions like ~1.2.3 and ^1.2.3

rsp opened the pull request:
It improves performance by downloading only the data of the version that is requested, instead of downloading data about every published version in the history of the package only to throw most of it away, as was the case before.

The difference is significant, for some modules like express it is 2kB vs 500kB, see:

time curl -s https://registry.npmjs.org/express | wc -c
time curl -s https://registry.npmjs.org/express/4.0.0 | wc -c

This is not only a waste of time and bandwidth from the perspective of the module user but also a significant waste of resources of the npm registry.

This PR also adds support for version numbers like:

* `1` - get the latest `1.x.x`
* `1.2` - get the latest `1.2.x`
* `^1.2.3` - get the latest `1.x.x` but at least `1.2.3`
* `~1.2.3` - get the latest `1.2.x` but at least `1.2.3`

so this module could be used to test most versions found in package.json dependencies.

It adds tests for the new version formats and makes sure that the data returned after this pull request is still the same as before.

It adds examples to the readme.

destroyerofbuilds commented on index.js:
```diff
@@ -4,9 +4,11 @@ var registryUrl = require('registry-url');
 	var url = registryUrl(scope) +
-		encodeURIComponent(name).replace(/^%40/, '@');
+		encodeURIComponent(name).replace(/^%40/, '@') +
+		(!isScoped && version ? '/' + version : '');
```
Though I agree this is more efficient, please read issue #7.

They are deprecating the version URL.

rsp replied:
@destroyerofbuilds If that is the case (which would be a very strange decision for a good API, I have to admit) then I guess the PR shouldn't be merged. Thanks for pointing it out. Do you maybe have some idea to get the real version number that would be installed by npm having a string like "~1.2.3" that would use the deprecated API that currently supports such version as part of the URL? Thanks.

sindresorhus replied:
You could use the [semver](https://github.com/npm/node-semver) module to match against the list of versions returned.

rsp replied:
@sindresorhus I'm using semver module in the PR #17, see: https://github.com/sindresorhus/package-json/pull/17/files

destroyerofbuilds commented on readme.md:
```diff
@@ -33,8 +33,16 @@ packageJson('@company/package', 'latest').then(json => {
+* `^1.2.3` - get the latest `1.x.x` but at least `1.2.3`
+* `~1.2.3` - get the latest `1.2.x` but at least `1.2.3`
+
+(Currently only full version numbers are supported for scoped packages.)
```
In these situations, could you link to supporting documentation as to why scoped packages are supported?

rsp replied:
Why they are supported, I have no idea, but apparently they are since you are already using them. What I meant by that was that scoped packages work after the PR just as they used to work before the PR, i.e. they still need a full version number. In other words, the PR doesn't change the behavior of the module when working with scoped packages.

destroyerofbuilds commented:
:-1: Because this approach depends on an API end-point that the npm team has indicated may be removed.

rsp commented:
@destroyerofbuilds Thanks for your comments. I prepared a new PR #17 that addresses all your concerns. It adds support for versions like `^1.2.3` and `~1.2.3` for both scoped and unscoped packages while still downloading the entire data for all versions in the package history every time.

sindresorhus commented:
Wish I could merge this, but don't want to depend on an API the owners has explicitly said will go away at some point.

I hope you can add a comment on https://github.com/npm/public-api/issues/11, though. It's really needed.

rsp commented:
@sindresorhus OK, I understand. @destroyerofbuilds pointed it out before so I created PR #17 that adds support for versions like ^1.2.3 and ~1.2.3 but without using that API - using the semver module instead. PR #17 adds the functionality to search for wildcard versions and ranges but doesn;t improve performance like using the single version API would - e.g. for express you download 500kB just to throweverything but 2kB immediately, see:
```
time curl -s https://registry.npmjs.org/express | wc -c
time curl -s https://registry.npmjs.org/express/4.0.0 | wc -c
```
It may not seem like a huge problem for some people but it will only get worse in my opinion with every new version.
